common/network-common/src/main/java/org/apache/spark/network/ssl/SSLFactory.java

```
@@ -0,0 +1,503 @@
+     * @param trustStoreReloadIntervalMs The interval at which to reload the trust store file
+     * @return The builder object
+     */
+    public Builder trustStore(
+        File trustStore, String trustStorePassword,
+        boolean trustStoreReloadingEnabled, int trustStoreReloadIntervalMs) {
+      this.trustStore = trustStore;
+      this.trustStorePassword = trustStorePassword;
+      this.trustStoreReloadingEnabled = trustStoreReloadingEnabled;
+      this.trustStoreReloadIntervalMs = trustStoreReloadIntervalMs;
+      return this;
+    }
+
+    /**
+     * Builds our {@link SSLFactory}
+     *
+     * @return The built {@link SSLFactory}
+     */
+    public SSLFactory build() {
+      return new SSLFactory(this);
+    }
+  }
+
+  /**
+   * Returns an initialized {@link SSLContext}
+   *
+   * @param requestedProtocol The requested protocol to use
+   * @param keyManagers The list of key managers to use
+   * @param trustManagers The list of trust managers to use
+   * @return The built {@link SSLContext}
+   * @throws GeneralSecurityException
+   */
+  private static SSLContext createSSLContext(
+      String requestedProtocol,
+      KeyManager[] keyManagers,
+      TrustManager[] trustManagers) throws GeneralSecurityException {
+    SSLContext sslContext = getSSLContextInstance(requestedProtocol);
+    sslContext.init(keyManagers, trustManagers, null);
+    return sslContext;
+  }
+
+  /**
+   * Get the {@link SSLContext} for the specified <tt>requestedProtocol</tt>
+   * if available, or the default {@link SSLContext}
+   * @param requestedProtocol The protocol to use
+   * @return The built {@link SSLContext}
+   * @throws NoSuchAlgorithmException
+   */
+  private static SSLContext getSSLContextInstance(String requestedProtocol)
+      throws NoSuchAlgorithmException {
+    SSLContext context = null;
+    try {
+      context = SSLContext.getInstance(requestedProtocol);
+    } catch (Exception e) {
+      context = SSLContext.getDefault();
```

Discuss: If the request protocol is invalid or unsupported, should we propagate the exception instead ?

---

adding a note: I was worried about not catching the `null` case when the user doesn't provide a protocol but I remembered the builder handles this. So we should propagate the exception and I'll make the code do that.